Profiling: intervals tool and devtools-file generator

kazcw opened the pull request:
### Pull Request Description

@MichaelMauderer and I decided we could use some more tooling:
- A tool that produces a quantitative report on total time spent in different profilers. (Hard numbers!)
- A tool that translates our profile format to the Chrome DevTools format. (This will give us a workaround until #3351 is fixed, and also has niceties we might not implement soon, like showing all labels and once, and showing durations on the chart.)

### Important Notes

- Both tools are run from the command line, and translate stdin to stdout.

### Checklist

Please include the following checklist in your PR:

- [x] The documentation has been updated if necessary.
- [x] All code conforms to the [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md), [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md), and [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guides.
- All code has been tested:
  - [x] Unit tests have been written where possible.
  - [x] If GUI codebase was changed: Enso GUI was tested when built using BOTH `./run dist` and `./run watch`.

[ci no changelog needed] internal tools

kazcw committed:
Profiling: intervals tool

kazcw committed:
devtools profile generator

kazcw commented:
`profiler` data in DevTools: 
![image](https://user-images.githubusercontent.com/1047859/159735602-3f371523-2cf6-44b3-8c6f-18cb2363e732.png)

kazcw commented:
Aggregated interval reports (how much time we are spending in each function, total):

```console
kaz@canoe:repo/enso/lib/rust/profiler/data$ cargo run --bin intervals < ~/profile.json
self_duration total_duration profiler
 535.5  535.5 parse_with_metadata (app/gui/language/parser/src/lib.rs:107)
 429.2  429.2 check (lib/rust/ensogl/core/src/system/gpu/shader.rs:96)
 365.8  365.8 build_and_mangle_glsl (lib/rust/ensogl/core/src/display/shape/primitive/shader/builder.rs:58)
 347.3  347.3 use_program (lib/rust/ensogl/core/src/display/symbol/gpu.rs:648)
 241.8  241.8 render_by_ids (lib/rust/ensogl/core/src/display/symbol/gpu/registry.rs:127)
 203.5  268.5 new (app/gui/src/controller/graph.rs:273)
 166.6  437.0 partial_dfs_with_layer_data (app/gui/language/span-tree/src/node.rs:760)
 164.9 1980.6 EventLoopExecutor::runner (app/gui/src/executor/web.rs:67)
 122.6  122.6 new (lib/rust/ensogl/core/src/animation/physics/inertia.rs:648)
  79.6  294.3 create_node (app/gui/view/graph-editor/src/lib.rs:1425)
  53.1   53.1 new (app/gui/src/controller/graph.rs:225)
  52.6   53.3 animation::physics::inertia::step (lib/rust/ensogl/core/src/animation/physics/inertia.rs:799)
  38.0  912.0 update (lib/rust/ensogl/core/src/display/scene.rs:1183)
  33.3   33.3 notify_language_server (app/gui/src/model/module/synchronized.rs:431)
  30.4   30.4 run (lib/rust/ensogl/core/src/display/render/passes/pixel_read.rs:163)
  28.9   43.8 set_node_expression_usage_type (app/gui/view/graph-editor/src/lib.rs:2075)
  26.9  497.6 update (/local/repo/enso/profiling-application/lib/rust/shapely/src/shared.rs:124)
  25.8 1235.2 go_to_next_frame_with_time (lib/rust/ensogl/core/src/display/world.rs:287)
  25.2  526.0 render_layer (lib/rust/ensogl/core/src/display/render/passes/symbols.rs:124)
  23.3   34.3 partial_dfs_with_layer_data (app/gui/language/span-tree/src/node.rs:595)
  22.7   22.7 set_node_expressions (app/gui/src/presenter/graph.rs:385)
  21.1  395.6 run (lib/rust/ensogl/core/src/display/shape/primitive/shader/builder.rs:38)
  16.2   27.6 nodes (app/gui/src/controller/graph.rs:548)
  15.8   15.8 nodes (app/gui/controller/double-representation/src/graph.rs:74)
  15.7  444.9 compile_shader (lib/rust/ensogl/core/src/system/gpu/shader.rs:135)
  15.6  470.3 set_node_expression (app/gui/view/graph-editor/src/lib.rs:1883)
  12.7  575.9 init (app/gui/src/presenter/graph.rs:476)
  11.3   21.2 start (app/gui/src/ide/initializer.rs:76)
  10.6  141.6 init_port_frp_on_new_expression (app/gui/view/graph-editor/src/component/node/input/area.rs:728)
  10.0   10.0 with_input_edge_id (app/gui/view/graph-editor/src/lib.rs:2222)
   9.9  365.2 set_pointer_events (lib/rust/ensogl/core/src/display/shape/primitive/system.rs:89)
   9.6 3278.7 @on_frame (lib/rust/ensogl/core/src/animation/loops.rs:161)
   9.2   12.4 new (app/gui/view/graph-editor/src/lib.rs:1645)
   8.7   11.0 draw (lib/rust/ensogl/core/src/display/shape/primitive/def/class.rs:40)
   8.1    8.1 redraw (lib/rust/ensogl/component/text/src/component/area.rs:731)
   7.9   42.2 init_new_expression (app/gui/view/graph-editor/src/component/node/input/area.rs:879)
   7.6   53.4 create_edge (app/gui/view/graph-editor/src/lib.rs:1335)
   7.5    7.5 with_output_edge_id (app/gui/view/graph-editor/src/lib.rs:2246)
   7.5    7.5 get_span_tree_node (app/gui/src/controller/graph.rs:238)
   7.0  407.4 set_expression (app/gui/view/graph-editor/src/component/node/input/area.rs:888)
   6.4  287.2 new (app/gui/src/presenter/graph.rs:361)
   5.6  503.2 update (/local/repo/enso/profiling-application/lib/rust/shapely/src/shared.rs:157)
   5.6    5.6 init_uniform_binding (lib/rust/ensogl/core/src/display/symbol/gpu.rs:561)
   5.5    5.5 set_projects_list_on_welcome_screen (app/gui/src/presenter.rs:222)
   5.4    5.4 handle_value_update (app/gui/src/presenter/graph/visualization.rs:100)
   4.5    4.5 build (lib/rust/ensogl/core/src/display/symbol/gpu/shader/builder.rs:362)
```

kazcw committed:
docs

kazcw committed:
Merge branch 'develop' into wip/kw/profiling-tools

kazcw commented:
A breakdown of our top 10, by self-time (time spent running a profiler, excluding time running any child profiler). This data is produced with my [deep-profiiing](https://github.com/enso-org/enso/tree/wip/kw/deep-profiling) branch, which produces an outlandish amount of data. It also *does not render text areas*, since the present implementation distorts the data by creating way too many shaders and @wdanilo is planning optimization to fix that.

` 535.5  535.5 parse_with_metadata (app/gui/language/parser/src/lib.rs:107)`
:hourglass_flowing_sand: [@wdanilo is working on this.](https://www.pivotaltracker.com/story/show/181553118)

` 429.2  429.2 check (lib/rust/ensogl/core/src/system/gpu/shader.rs:96)`
:bangbang: We are spending all this time blocking on something done in the background! [See 2nd para in comment.](https://www.pivotaltracker.com/story/show/181632388/comments/230287801)

` 365.8  365.8 build_and_mangle_glsl (lib/rust/ensogl/core/src/display/shape/primitive/shader/builder.rs:58)`
:bangbang: We are doing something I'm not sure should be necessary, really inefficiently! [See 1st para in comment](https://www.pivotaltracker.com/story/show/181632388/comments/230287801).

` 347.3  347.3 use_program (lib/rust/ensogl/core/src/display/symbol/gpu.rs:648)`
:zzz: Switching shaders is costly. I believe we already take the necessary steps to do it as little as possible, though I have not verified there isn't room for improvement here.

` 241.8  241.8 render_by_ids (lib/rust/ensogl/core/src/display/symbol/gpu/registry.rs:127)`
:zzz: Actual rendering; here too, we can reasonably expect to spend a lot of time.

` 203.5  268.5 new (app/gui/src/controller/graph.rs:273)`
🔍 Probably improvable, but no obvious huge inefficiencies.

` 164.9 1980.6 EventLoopExecutor::runner (app/gui/src/executor/web.rs:67)`
:zzz: All not-yet-identified work done in async tasks. Most likely a sum of misc stuff that aren't individually significant.

` 149.2  150.2 start (lib/rust/ensogl/core/src/animation/physics/inertia.rs:706)`
:exclamation: Starting animation loops. We should dig further into this; I suspect potential for optimization here.

` 103.3  109.4 code_coloring (app/gui/view/graph-editor/src/component/node/input/area.rs:764)`
🔍 Initializing FRP networks for graph nodes. I'd expect that to have some cost, but we could look into microoptimizing it after we run out of easy improvements.

`  79.6  294.3 create_node (app/gui/view/graph-editor/src/lib.rs:1425)`
🔍 Probably improvable, but no obvious huge inefficiencies.

kazcw committed:
Merge branch 'develop' into wip/kw/profiling-tools

MichaelMauderer commented on lib/rust/profiler/data/src/bin/devtools.rs:
```diff
@@ -0,0 +1,130 @@
+        let active = &self.profile[active];
+        let measurement = &self.profile[active.measurement];
+        let start = active.interval.start.into_ms();
+        // TODO[kw]: The format supports incomplete events, but isn't documented.
```
Can this be easily solved as part of this PR or is it much more work?

kazcw replied:
I just tested that, and it turns out DevTools ignores open intervals; I changed the implementation to skip emitting them.

kazcw committed:
ignore open intervals, because devtools does

kazcw committed:
Merge branch 'develop' into wip/kw/profiling-tools

kazcw committed:
intervals tool: show interval counts

kazcw commented:
@wdanilo I added interval-counts to the tool's output:
```console
self_duration total_duration count profiler
 528.7  528.7 1 parse_with_metadata (app/gui/language/parser/src/lib.rs:107)
 390.5  390.5 356 check (lib/rust/ensogl/core/src/system/gpu/shader.rs:96)
 383.6  383.6 406 build_and_mangle_glsl (lib/rust/ensogl/core/src/display/shape/primitive/shader/builder.rs:58)
 361.1  361.1 177 use_program (lib/rust/ensogl/core/src/display/symbol/gpu.rs:648)
 220.3  220.3 1416 render_by_ids (lib/rust/ensogl/core/src/display/symbol/gpu/registry.rs:127)
 190.6  250.7 255 new (app/gui/src/controller/graph.rs:273)
 159.5 1985.8 134 EventLoopExecutor::runner (app/gui/src/executor/web.rs:67)
 149.2  150.2 1231 start (lib/rust/ensogl/core/src/animation/physics/inertia.rs:706)
 103.3  109.4 312 code_coloring (app/gui/view/graph-editor/src/component/node/input/area.rs:764)
  74.8  303.9 17 create_node (app/gui/view/graph-editor/src/lib.rs:1425)
```

We are building a lot of shaders. In particular, we are building a lot more than we are using!

I'm going to look in to whether this is downstream of text-area work that's happening outside of `redraw`.

kazcw commented:
Shaders built: 406 (build_and_mangle_glsl)
Shaders compiled: 178 (356 checks, vertex+fragment)
use_program: 177

I don't know how many times we're using each program, so I can't tell if we're using most of the shaders we're compiling. But we're definitely not compiling most of the shaders we build, which I suspect is caused by creating text areas and never drawing. So, likely about 1/2 the time in build_and_mangle_glsl is for text-area inefficiencies. I'm working on confirming. In any case, I don't think this will change the low-hanging fruit here.

kazcw committed:
Merge branch 'develop' into wip/kw/profiling-tools

kazcw commented:
Ok, I further disabled text-area so that it never even loads a font.

Workflow: open to welcome screen -> create project

```console
self_duration total_duration count profiler
 528.6  528.6 1 parse_with_metadata (app/gui/language/parser/src/lib.rs:107)
 345.4  345.4 33 use_program (lib/rust/ensogl/core/src/display/symbol/gpu.rs:648)
 308.9  308.9 66 check (lib/rust/ensogl/core/src/system/gpu/shader.rs:96)
 243.7  243.7 265 build_and_mangle_glsl (lib/rust/ensogl/core/src/display/shape/primitive/shader/builder.rs:58)
 200.0  264.6 255 new (app/gui/src/controller/graph.rs:273)
 160.4 1841.0 134 EventLoopExecutor::runner (app/gui/src/executor/web.rs:67)
 151.4  151.4 1272 new (lib/rust/ensogl/core/src/animation/loops.rs:67)
  98.4  103.2 315 code_coloring (app/gui/view/graph-editor/src/component/node/input/area.rs:764)
  72.3   72.3 700 render_by_ids (lib/rust/ensogl/core/src/display/symbol/gpu/registry.rs:127)
  68.1  180.0 17 create_node (app/gui/view/graph-editor/src/lib.rs:1425)
```

This makes a big difference in how many times the different GPU-related functions are called, but a more moderate difference to how much time we spend in check/build_and_mangle (I guess on average, the non-text-area symbols are more expensive to draw than the text-area symbols.). So this more accurate data probably doesn't change our priorities; `use_program` is now relatively more important, but is not likely optimizable.

Workflow: open directly to a project with '--project'

```console
self_duration total_duration count profiler
 542.3  542.3 1 parse_with_metadata (app/gui/language/parser/src/lib.rs:107)
 326.9  326.9 32 use_program (lib/rust/ensogl/core/src/display/symbol/gpu.rs:648)
 300.3  300.3 64 check (lib/rust/ensogl/core/src/system/gpu/shader.rs:96)
 239.0  239.0 256 build_and_mangle_glsl (lib/rust/ensogl/core/src/display/shape/primitive/shader/builder.rs:58)
 193.9  259.7 255 new (app/gui/src/controller/graph.rs:273)
 151.1 1869.3 126 EventLoopExecutor::runner (app/gui/src/executor/web.rs:67)
 143.4  143.4 1267 new (lib/rust/ensogl/core/src/animation/loops.rs:67)
  99.0  107.3 322 code_coloring (app/gui/view/graph-editor/src/component/node/input/area.rs:764)
  81.1  211.2 17 create_node (app/gui/view/graph-editor/src/lib.rs:1425)
  59.1   59.1 575 render_by_ids (lib/rust/ensogl/core/src/display/symbol/gpu/registry.rs:127)
```

kazcw committed:
Merge branch 'develop' into wip/kw/profiling-tools

kazcw committed:
Re-add a doc comment. Was accidentally deleted.

wdanilo commented on lib/rust/profiler/data/src/aggregate.rs:
```diff
@@ -0,0 +1,87 @@
+
+use std::collections;
+use std::rc;
+
+/// Aggregate time spent in different functions.
+#[derive(Default, Debug)]
+pub struct Aggregator {
+    stack: Vec<std::rc::Rc<String>>,
+    root:  Frame,
+}
+
```
no sections. Please remember that each file should contain at least one section.

wdanilo commented on lib/rust/profiler/data/src/aggregate.rs:
```diff
@@ -0,0 +1,87 @@
+}
+
+impl Aggregator {
+    /// Add data from a profile to the tree.
+    pub fn visit_profile<Metadata>(&mut self, profile: &crate::Profile<Metadata>) {
```
I'm confused why `visit_profile` is "addicting data from a profile to the tree". The name suggests to me that this is a standard visitor impl, but it looks it does something else. Do you think we could improve the naming here and in associated functions?

kazcw replied:
Renamed the functions that apply a profile to the aggregated data `add_profile`.

wdanilo commented on lib/rust/profiler/data/src/aggregate.rs:
```diff
@@ -0,0 +1,87 @@
+    }
+}
+
+/// Aggregated info about all occurrences of a particular stack of profilers.
```
missing another section

wdanilo commented on lib/rust/profiler/data/src/aggregate.rs:
```diff
@@ -0,0 +1,87 @@
+pub struct Frame {
+    duration:     f64,
+    /// Aggregated intervals that ran as children of this profiler.
+    pub children: collections::HashMap<rc::Rc<String>, Self>,
```
Instead of `Rc<String>` you can use `ImString` from `Prelude`.

wdanilo commented on lib/rust/profiler/data/src/bin/devtools.rs:
````diff
@@ -0,0 +1,131 @@
+//! Tool that generates Chrome DevTools-compatible files from profiling interval data.
+//!
+//! # Usage
+//!
+//! The tool reads a JSON-formatted event log from stdin, and writes a report to stdout.
+//!
+//! For example:
+//!
+//! ```console
+//! ~/git/enso/data $ cargo run --bin intervals < profile.json > devtools.json
+//! ```
````
These docs don't explain several important things:
1. Where is the spec this code bases on? Where are the descriptions of the format supported by Chrome, so we can easily refer to it in the future when extending this code? (ofc this can be provided in other docs than module docs, but I haunt found it below).
2. What is `profile.json`. I know the context so I understand it, but someone reading this code might not. Such docs should explain the context, including short info of what profile.json is, how to generate it, and should provide step-by step guide of how to get it running starting with 0 knowledge.

kazcw replied:
1. Official documentation is unavailable. I've added a link to a Google Doc I found that seems to be someone's reverse-engineering of the format.
2. I've added a link to the appropriate section of the design doc. The design doc is out of date wrt. the procedure for capturing the data since we decided to implement an interactive workflow before we implement batch mode; I've opened a [PR for the design doc](https://github.com/enso-org/design/pull/30) to describe the capture procedure as implemented. Also, I've added a commit here to implement the `--profiling-level` build script switch described in the design doc.

wdanilo commented on lib/rust/profiler/data/src/bin/intervals.rs:
````diff
@@ -0,0 +1,101 @@
+//! Tool that generates interval reports from profiling data.
+//!
+//! # Usage
+//!
+//! The tool reads a JSON-formatted event log from stdin, and writes a report to stdout.
+//!
+//! For example:
+//!
+//! ```console
+//! ~/git/enso/data $ cargo run --bin intervals < profile.json | less
+//! ```
````
I have here the same comment as to the `devtools.rs` file – I'd be very thankful if you describe the context more, so people that are new to the codebase have easy time understanding what are all the components here, including what `profile.json` is and how to get it (this can, ofc, be a redirection to another docs).

wdanilo commented on lib/rust/profiler/data/src/bin/intervals.rs:
```diff
@@ -0,0 +1,101 @@
+    let root = data::aggregate::Frame::from(aggregator);
+    let funcs = FuncCollector::run(&root);
+    let mut funcs: Vec<_> = funcs.into_iter().collect();
+    funcs.sort_by_key(|(_, func)| 0 - (func.self_duration * 10.0).round() as i64);
```
I'd love to see either docs or refactoring to a function/lambda this equation, because it is not clear to me why it works this way.

kazcw replied:
You're right, that's completely opaque. I've rewritten it.

wdanilo commented on lib/rust/profiler/data/src/bin/intervals.rs:
```diff
@@ -0,0 +1,101 @@
+    }
+}
+
+/// Aggregate of all time spent in a particular profiler's intervals.
+#[derive(Default)]
+struct Func {
+    total_duration: f64,
+    self_duration:  f64,
+    count:          usize,
+}
+
+type Label = std::rc::Rc<String>;
```
This is defined in the `FuncCollector` section – is this correct?

wdanilo commented on lib/rust/profiler/flame-graph/src/lib.rs:
```diff
@@ -163,7 +163,7 @@ impl<'p, Metadata> RungraphBuilder<'p, Metadata> {
                 let start = active.interval.start.into_ms();
                 let mut end = active.interval.end.map(|mark| mark.into_ms()).unwrap_or(f64::MAX);
                 // We want to show every wakeup of the task, even if it is momentary.
-                const DURATION_FLOOR_MS: f64 = 0.5;
+                const DURATION_FLOOR_MS: f64 = 2.0;
```
I missed that in previous review – these docs should explain a little bit better what `DURATION_FLOOR_MS` is. The explanation "We want to show every wakeup of the task, even if it is momentary." is not clear IMO. If I understand correctly, this is a minimal "width" of a bar in the flame graph to make it easier visible, right?

kazcw replied:
Yes, but it's in the run graph, which illustrates async relationships. (In the flame graph, short intervals don't matter.) Extended the docs.

kazcw committed:
Implement review.

kazcw committed:
DURATION_FLOOR_MS

kazcw committed:
Rename methods that apply a profile's contents to self

kazcw committed:
Description of chrome format

kazcw committed:
Link to profile.json format documentation is design doc

kazcw committed:
Implement profiling-level switch described in design doc

kazcw committed:
lint

kazcw commented:
@wdanilo I've implemented the review edits. I also added a commit that implements the `--profiling-level` flag for the build script, to match the design (I'd been using the lower-level environment-variable interface directly to get data, but that isn't supposed to be public). There's a companion docs PR: enso-org/design#30

kazcw committed:
Merge branch 'develop' into wip/kw/profiling-tools

kazcw committed:
Merge branch 'develop' into wip/kw/profiling-tools

kazcw committed:
Merge branch 'develop' into wip/kw/profiling-tools

MichaelMauderer commented:
@kazcw @wdanilo Is this still awaiting re-review? I would love to see this merged.

kazcw committed:
Merge branch 'develop' into wip/kw/profiling-tools

wdanilo commented:
Sorry, in case I'm not responding fast, ping me on Discord. This should be accepted long time ago.
